pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/model/BuildCondition.java

```diff
@@ -130,8 +140,28 @@ protected Result getExecutionResult(@Nonnull WorkflowRun r) {
         return conditions;
     }
 
+    @Deprecated
     @Nonnull
     public static Result getCombinedResult(@Nonnull WorkflowRun run, @CheckForNull Throwable error) {
+        return getCombinedResult(run, error, null, null);
+    }
+
+    @Nonnull
+    public static Result getCombinedResult(@Nonnull WorkflowRun run, @CheckForNull Throwable error, @CheckForNull FlowNode startNode, @CheckForNull FlowNode endNode) {
```

Comment:
FWIW I think you could remove some duplicated code by changing this to `getCombinedResult(@Nonnull WorkflowRun run, @CheckForNull Throwable error, @CheckForNull String stageName)` and doing the following here rather than in all the specific-result conditions:

```
FlowNode startNode = null
FlowNode endNode = null
if (context instanceof Stage) {
    (startNode, endNode) = Utils.findPossiblyUnfinishedEndNodeForCurrentStage(((Stage)context).name, r.getExecution())
}
```

Reply:
...if I could, I would, but `Stage` is in `pipeline-model-definition`, while this is in `pipeline-model-extensions`...

Reply:
But you just need the name, right? You can do the `(Stage)context` cast up in the conditions and just pass down the String, although I guess you'd need to move `findPossiblyUnfinishedEndNodeForCurrentStage` to pipeline-model-extensions to make it work.

Reply:
Yeah. I have regrets about the organization. =)

Reply:
Yeah, my main thought is that if https://github.com/jenkinsci/pipeline-model-definition-plugin/pull/330#discussion_r290757003 is a problem, you'll need the name of the stage wherever your `findWarningBetween` logic lives to make it able to skip nodes in nested stages, so since `Util.findPossiblyUnfinishedEndNodeForCurrentStage` and `Util.isSomewhereWithinStage` are new methods anyway, might as well move them all into pipeline-model-extensions and just pass the stage name to `BuildCondition.whatever` and then do all of the work in one `findWorstWarningActionForStage(String stageName)` method over there. It isn't quite that clean though because you'd have to dupe or move `Util.isStageWithOptionalName`. If that turns out not to be a problem, then it doesn't matter.

Reply:
Hmm. Lemme think on that.